Adds support for Carthage

bdorfman opened the pull request:
This PR adds Carthage support to PINRemoteImage. It also as a side effect should make it slightly easier for people who want to manually clone/submodule the repo and build the framework to do so. However, I am concerned I may have accidentally broke the existing Cocoapods support since I don't quite understand the Cocoapods flow, so there may be further changes needed in that regard before this can be merged.

This adds a new Xcode project that builds PINRemoteImage.framework. Xcode wants frameworks' umbrella/master header file (the one named FrameworkName.h) to import all public headers in the framework or it throws a warning. To accomplish this I had to move macros out of PINRemoteImage.h into a separate file to resolve circular dependency issues. I also properly guarded the compilation of FLAnimatedImage code around whether or not the library is available, and removed the OS_IPHONE check as this framework does not support OS X anyway (and in order to do so would need a lot more changes than just swapping UIImage for NSImage).

PINCache is added as a Carthage dependency so if you use Carthage that will be taken care of for you automatically. It should theoretically also work with FLAnimatedImage via Carthage if you have that in your parent app's Cartfile, although their Carthage support seems a little broken at the moment so I couldn't test it properly.

garrettmoon commented:
Sorry I haven't looked at this yet, I'll try and take a look next week.

bdorfman commented:
No worries, I can just use my own fork for our app so there's no rush.

garrettmoon commented on PINRemoteImage/PINRemoteImage.xcodeproj/project.xcworkspace/xcshareddata/PINRemoteImage.xcscmblueprint:
```diff
@@ -0,0 +1,30 @@
+      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryIdentifierKey" : "44B36422A7A7C3AD3DE414BC27AA89E3C0009BBA"
+    },
+    {
+      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "github.com:foursquare\/PINRemoteImage.git",
```
Should this file be checked in? Should anything in xcshareddata?

bdorfman replied:
Ah yeah, that is probably a mistake. You need to check in your shared framework scheme (which is in xcshareddata) so that Carthage can find and build it, but not this blueprint file I don't think. I will fix.

garrettmoon commented:
Thank you so much for doing this! I tested it locally and it seems CocoaPods is still happy :) Going to do a bit more due diligence and then I think we can get this merged!

bdorfman commented:
No problem. By the way, I learned more about modulemap minutiae since doing this, so I think I can convince it to not require that all public headers be imported by PINCache.h if you would like me to change that back.

FYI, proper support for FLAnimatedImage via Carthage requires them to merge this PR https://github.com/Flipboard/FLAnimatedImage/pull/95 But theoretically if you list FLAnimatedImage in your Cartfile before you list PINRemoteImage, it will build it with animated gif support.

garrettmoon commented:
If we can at least get the category imports out that might be a good idea?

bdorfman commented:
Ok, I think everything should work now. Updated gitignore, removed unnecessary xcode files, update the header/modulemap, and added carthage instructions to readme.

I think the travis-ci build is failing because I don't know how to update the example project properly using cocoapods. I tried just doing `pod install` and that changed a LOT of files that I wasn't sure about so I will just leave it up to you to resolve the proper way.

garrettmoon commented:
Cool! I'll get this merged.

garrettmoon commented:
Manually merged, thanks so much for doing this!

garrettmoon commented:
Hey @bdorfman just tried to figure out this whole Carthage thing and I'm receiving the same error as here:

https://github.com/pinterest/PINRemoteImage/issues/25#issuecomment-153346771

Should there be an xcodeproj specifically for Carthage in the root directory with a shared framework scheme? Did it not get checked in?
